File: 8 - Memoization Search & Dynamic Programming/E828.WordPattern.py
class Solution:
    """
    @param pattern: a string, denote pattern string
    @param teststr: a string, denote matching string
    @return: an boolean, denote whether the pattern string and the matching string match or not
    """
    def wordPattern(self, pattern, teststr):
        # write your code here

        words = teststr.split()

        words_size, hashmap = len(words), {}
        if words_size != len(pattern): return False

        for i, word in enumerate(words):
            if pattern[i] not in hashmap:
                hashmap[pattern[i]] = word
            elif hashmap[pattern[i]] != word:
                return False

        # make sure the keys count is equal to the values count
        if len(hashmap) != len(set(hashmap.values())): return False

        return True

File: 08 - Memoization Search & Dynamic Programming/E828.WordPattern.java
public class Solution {
    /**
     * @param pattern: a string, denote pattern string
     * @param teststr: a string, denote matching string
     * @return: an boolean, denote whether the pattern string and the matching string match or not
     */
    public boolean wordPattern(String pattern, String teststr) {
        // write your code here
        String[] words = teststr.split(" ");

        if (words.length != pattern.length()) return false;

        Map<Character, String> hashmap = new HashMap<Character, String>();
        for (int i = 0; i < words.length; i++) {
            if (!hashmap.containsKey(pattern.charAt(i))) {
                if (hashmap.containsValue(words[i])) return false;
                hashmap.put(pattern.charAt(i), words[i]);
            } else if (!words[i].equals(hashmap.get(pattern.charAt(i))))
                return false;
        }

        return true;
    }
}
